File: payment_mcp/signature.py
import hashlib
from urllib.parse import unquote


def build_signing_payload(data: dict[str, object], sign_key: str) -> str:
    normalized = {}
    for key, value in data.items():
        if key == "sign" or value is None:
            continue
        raw = str(value)
        if key == "notify_url":
            raw = unquote(raw)
        normalized[key] = raw

    pairs = [f"{key}={normalized[key]}" for key in sorted(normalized.keys())]
    return f"{'&'.join(pairs)}&signkey={sign_key}"


def md5_sign(data: dict[str, object], sign_key: str) -> str:
    payload = build_signing_payload(data, sign_key)
    return hashlib.md5(payload.encode("utf-8")).hexdigest()

File: node-mcp/src/signature.ts
import { createHash } from "node:crypto";

type SignableValue = string | number | boolean | null | undefined;

export type SignablePayload = Record<string, SignableValue>;

export function buildSigningPayload(data: SignablePayload, signKey: string): string {
  const pairs = Object.entries(data)
    .filter(([key, value]) => key !== "sign" && value !== null && value !== undefined)
    .map(([key, value]) => {
      const rawValue = key === "notify_url" ? decodeURIComponent(String(value)) : String(value);
      return [key, rawValue] as const;
    })
    .sort(([left], [right]) => left.localeCompare(right))
    .map(([key, value]) => `${key}=${value}`);

  return `${pairs.join("&")}&signkey=${signKey}`;
}

export function md5Sign(data: SignablePayload, signKey: string): string {
  return createHash("md5").update(buildSigningPayload(data, signKey), "utf8").digest("hex");
}
